throttle.go

```diff
@@ -0,0 +1,55 @@
+package taskgroup
+
+// A Throttle rate-limits the number of concurrent goroutines that can execute
+// in parallel to some fixed number.  A zero Throttle is ready for use, but
+// imposes no limit on parallel execution. See [Throttle.Enter] for use.
+type Throttle struct {
+	adm chan struct{}
+}
+
+// NewThrottle constructs a [Throttle] with a capacity of n goroutines.
+// If n ≤ 0, the resulting Throttle imposes no limit.
+func NewThrottle(n int) Throttle {
+	if n <= 0 {
+		return Throttle{}
+	}
+	return Throttle{adm: make(chan struct{}, n)}
+}
+
+// Enter blocks until a slot is available in t, then returns a [Leaver] that
+// the caller must execute to return the slot when it is no longer in use.
+func (t Throttle) Enter() Leaver {
```

Comment:
This signature almost looks like context.WithCancel... I don't know if there's anything in that, just noticed and wondering if there's some opportunity to move the two closer together. Probably not, but sharing in case it causes any flash of insight.

Reply:
It kind of does, but I'm not sure I see the rhyme: `WithCancel` returns the new context and the CancelFunc, on the principle you ought to be using both. But here you kind of already have to have the Throttle in-hand. Am I missing something?